File: scripts/update_exercises.js
const fs = require('fs').promises;
const path = require('path');

async function updateExercises() {
  try {
    // Read content files
    const weekFiles = [
      'Week 1 Spanish.txt',
      'Week 2 Spanish.txt',
      'Week 3 Spanish.txt',
      'Week 4 Spanish.txt',
      'Week 5 Spanish.txt'
    ];

    const exercises = [];

    // Parse content files
    for (const file of weekFiles) {
      const content = await fs.readFile(path.join(__dirname, '../CONTENT', file), 'utf-8');
      const exercisesInFile = parseContent(content, file);
      exercises.push(...exercisesInFile);
    }

    // Update/create exercise files
    for (const exercise of exercises) {
      await createOrUpdateExerciseFile(exercise);
    }

    console.log('All exercises have been updated successfully!');
  } catch (error) {
    console.error('Error updating exercises:', error);
  }
}

function parseContent(content, fileName) {
  const exercises = [];
  const lines = content.split('\n');
  let currentExercise = null;
  let currentSection = null;
  let week = null;
  let day = null;

  // Extract week number from filename
  const weekMatch = fileName.match(/Week (\d+)/);
  if (weekMatch) {
    week = weekMatch[1];
  }

  let currentExerciseNumber = 1; // Track exercise number for fallback

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();

    // Match day header
    if (line.startsWith('Semana') && line.includes('Día')) {
      const dayMatch = line.match(/Semana (\d+) - Día (\d+)/);
      if (dayMatch) {
        day = dayMatch[2];
        currentExerciseNumber = 1; // Reset exercise number for new day
      }
    }

    // Skip the special case line
    if (line === 'Ejercicio de IR y extensión de cadera en decúbito prono') {
      continue;
    }

    // Match exercise header
    if (line.startsWith('Ejercicio')) {
      if (currentExercise) {
        exercises.push(currentExercise);
      }

      // Try to parse exercise number from title
      const match = line.match(/Ejercicio (\d+): (.+)/);
      if (match) {
        currentExerciseNumber = parseInt(match[1]);
        currentExercise = {
          title: line,
          setup: [],
          instructions: [],
          setsReps: [],
          whatToFeel: [],
          whatToAvoid: [],
          redFlags: [],
          regression: [],
          week,
          day
        };
      } else {
        // Fallback for invalid titles
        currentExercise = {
          title: `Ejercicio ${currentExerciseNumber}: ${line.replace('Ejercicio', '').trim()}`,
          setup: [],
          instructions: [],
          setsReps: [],
          whatToFeel: [],
          whatToAvoid: [],
          redFlags: [],
          regression: [],
          week,
          day
        };
      }
      currentSection = null;
    }

    // Match section headers
    if (currentExercise) {
      if (line.startsWith('Configuración:')) {
        currentSection = 'setup';
      } else if (line.startsWith('Instrucciones:')) {
        currentSection = 'instructions';
        currentExercise.instructions = []; // Clear instructions section
      } else if (line.startsWith('Conjuntos:')) {
        currentSection = 'setsReps';
      } else if (line.startsWith('Qué sentir:')) {
        currentSection = 'whatToFeel';
      } else if (line.startsWith('Lo que hay que evitar:')) {
        currentSection = 'whatToAvoid';
      } else if (line.startsWith('Banderas rojas:')) {
        currentSection = 'redFlags';
      } else if (line.startsWith('Regresión:')) {
        currentSection = 'regression';
      } else if (line.startsWith('Semana')) {
        // Stop parsing regression when we reach the Semana line
        if (currentSection === 'regression') {
          currentSection = null;
        }
      }

      // Add content to current section
      if (currentSection && line && !line.startsWith('Ejercicio')) {
        // Skip section headers for most sections
        if (currentSection !== 'setsReps' && 
            (line === 'Configuración:' || 
             line === 'Instrucciones:' || 
             line === 'Qué sentir:' || 
             line === 'Lo que hay que evitar:' || 
             line === 'Banderas rojas:' || 
             line === 'Regresión:')) {
          continue;
        }

        // Remove asterisks from content
        const cleanedLine = line.replace(/\*/g, '');

        // For regression, skip the header and line
        if (currentSection === 'regression' && cleanedLine !== 'Regresión:' && cleanedLine !== '________________') {
          currentExercise[currentSection].push(cleanedLine);
        } else if (currentSection === 'setsReps') {
          // Keep the labels for setsReps section and format them properly
          const parts = cleanedLine.split(':');
          if (parts.length === 2) {
            const label = parts[0].trim();
            const value = parts[1].trim();
            currentExercise[currentSection].push(`${label}: ${value}`);
          } else {
            currentExercise[currentSection].push(cleanedLine);
          }
        } else if (currentSection !== 'regression') {
          // Stop copying instructions when we encounter Semana
          if (currentSection === 'instructions' && cleanedLine.startsWith('Semana')) {
            break;
          }
          currentExercise[currentSection].push(cleanedLine);
        }
      }
    }
  }

  if (currentExercise) {
    exercises.push(currentExercise);
  }

  return exercises;
}

async function createOrUpdateExerciseFile(exercise) {
  // Extract exercise number from title
  const match = exercise.title.match(/Ejercicio (\d+): (.+)/);
  if (!match) {
    console.error('Invalid exercise title:', exercise.title);
    return;
  }

  const [_, number, title] = match;
  const exerciseNumber = parseInt(number);

  // Create function name
  const functionName = `Ejercicio${exerciseNumber}${title.replace(/[^a-zA-Z0-9]/g, '')}`;

  // Create file content
  const fileContent = `import EspDrillBarTemplate from '../../../../components/EspDrillBarTemplate';
import EspDrillNavBar from '../../../../components/EspDrillNavBar';
import VideoCard from '../../../../components/VideoCard';
import Link from 'next/link';

export default function ${functionName}() {
  return (
    <EspDrillBarTemplate
      title="${exercise.title}"
      backButtonText="Semana ${exercise.week} - Día ${exercise.day}"
      videoId="wn0IyvGBeUI"
      videoStart={2248}
      setup={${JSON.stringify(exercise.setup)}}
      instructions={${JSON.stringify(exercise.instructions)}}
      setsReps={${JSON.stringify(exercise.setsReps)}}
      whatToFeel={${JSON.stringify(exercise.whatToFeel)}}
      whatToAvoid={${JSON.stringify(exercise.whatToAvoid)}}
      redFlags={${JSON.stringify(exercise.redFlags)}}
      regression={${JSON.stringify(exercise.regression)}}
    >
      <EspDrillNavBar 
        prevDrill="/espanol/semana${exercise.week}/dia${exercise.day}/ejercicio${exerciseNumber - 1}" 
        currentDay="/espanol/semana${exercise.week}/dia${exercise.day}" 
        nextDrill="/espanol/semana${exercise.week}/dia${exercise.day}/ejercicio${exerciseNumber + 1}" 
        disablePrevious={${exerciseNumber === 1}}
        disableNext={${exerciseNumber === 5}}
      />
    </EspDrillBarTemplate>
  );
}
`;

  // Create file path
  const filePath = path.join(__dirname, '../pages/espanol/semana' + exercise.week + '/dia' + exercise.day + '/ejercicio' + exerciseNumber + '.js');

  // Create directories if they don't exist
  await fs.mkdir(path.dirname(filePath), { recursive: true });

  // Write file
  await fs.writeFile(filePath, fileContent, 'utf-8');
  console.log('Updated', filePath);
}

// Run the script
updateExercises();

File: scripts/update_exercises.py
import csv
import os
from pathlib import Path
from tqdm import tqdm
import re

class ExerciseUpdater:
    def __init__(self):
        self.base_path = Path("pages/espanol")
        self.csv_path = Path("RESOURCES/spanishdrills.csv")
        self.report = {
            'total_files': 0,
            'updated_files': 0,
            'errors': [],
            'unchanged': [],
            'mismatches': []
        }

    def load_csv(self):
        """Load and parse the CSV file."""
        try:
            with open(self.csv_path, 'r', encoding='utf-8') as f:
                reader = csv.DictReader(f)
                self.exercises = list(reader)
            return True
        except Exception as e:
            self.report['errors'].append(f"Error loading CSV: {str(e)}")
            return False

    def verify_file_structure(self):
        """Verify that all expected files exist."""
        expected_files = set()
        
        for ex in self.exercises:
            week = ex['Semana']
            day = ex['Día']
            number = ex['Ejercicio #']
            
            # Create full path with proper week/day numbering
            week_num = week.split()[1]  # Extract the number from "Semana X"
            day_num = day.split()[1]    # Extract the number from "Día X"
            
            file_path = self.base_path / f"semana{week_num}" / f"dia{day_num}" / f"ejercicio{number}.js"
            expected_files.add(file_path)
            
            if not file_path.exists():
                self.report['errors'].append(f"Missing file: {file_path}")
        
        self.report['total_files'] = len(expected_files)
        return len(self.report['errors']) == 0

    def update_file(self, file_path, exercise):
        """Update a single exercise file."""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()

            # Verify current content
            # Check for EspDrillBarTemplate title
            current_title = re.search(r'title="([^"]+)"', content)
            
            # Check for EspDrillNavBar
            current_nav = re.search(r'<EspDrillNavBar.*?>', content)
            
            if not (current_title and current_nav):
                self.report['errors'].append(f"Invalid format in {file_path}")
                return False

            # Check for duplicate description attributes in EspDrillNavBar
            desc_count = len(re.findall(r'description="([^"]+)"', current_nav.group(0)))
            if desc_count > 1:
                self.report['errors'].append(f"Duplicate description attributes in {file_path}")
                return False

            # Update EspDrillNavBar
            nav_bar_pattern = r'<EspDrillNavBar.*?>'
            
            # Get week and day numbers
            week_num = exercise['Semana'].split()[1]
            day_num = exercise['Día'].split()[1]
            
            # Calculate next exercise number
            next_num = int(exercise['Ejercicio #']) + 1
            
            # Create proper paths
            next_drill_path = f"/espanol/semana{week_num}/dia{day_num}/ejercicio{next_num}"
            current_day_path = f"/espanol/semana{week_num}/dia{day_num}"
            
            # Create updated nav bar
            nav_bar_update = f'''
      <EspDrillNavBar 
        currentDay="{current_day_path}" 
        nextDrill="{next_drill_path}" 
        disablePrevious={{ {exercise['Ejercicio #'] == '1'} }} 
        disableNext={{{False}}}
      />
    '''
            
            # Replace the old nav bar with the new one
            updated_content = re.sub(nav_bar_pattern, nav_bar_update, content)
            
            # Write changes back to file
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(updated_content)
                self.report['updated_files'] += 1

            return True

            # Check for mismatches
            if current_title.group(1) != exercise['Ejercicio']:
                self.report['mismatches'].append({
                    'file': str(file_path),
                    'field': 'title',
                    'current': current_title.group(1),
                    'expected': exercise['Ejercicio']
                })

            if current_desc.group(1) != exercise['Descripción']:
                self.report['mismatches'].append({
                    'file': str(file_path),
                    'field': 'description',
                    'current': current_desc.group(1),
                    'expected': exercise['Descripción']
                })

            # Check for mismatches
            if current_title.group(1) != exercise['Ejercicio']:
                self.report['mismatches'].append({
                    'file': str(file_path),
                    'field': 'title',
                    'current': current_title.group(1),
                    'expected': exercise['Ejercicio']
                })

            if current_desc.group(1) != exercise['Descripción']:
                self.report['mismatches'].append({
                    'file': str(file_path),
                    'field': 'description',
                    'current': current_desc.group(1),
                    'expected': exercise['Descripción']
                })

            # Update content
            updated_content = content
            updated_content = re.sub(r'title="([^"]+)"', f'title="{exercise["Ejercicio"]}"', updated_content)
            updated_content = re.sub(r'description="([^"]+)"', f'description="{exercise["Descripción"]}"', updated_content)

            # Update EspDrillNavBar
            nav_bar_pattern = r'<EspDrillNavBar.*?>'
            
            # Get week and day numbers
            week_num = exercise['Semana'].split()[1]
            day_num = exercise['Día'].split()[1]
            
            # Calculate next exercise number
            next_num = int(exercise['Ejercicio #']) + 1
            
            # Create proper paths
            next_drill_path = f"/espanol/semana{week_num}/dia{day_num}/ejercicio{next_num}"
            current_day_path = f"/espanol/semana{week_num}/dia{day_num}"
            
            # Create updated nav bar
            nav_bar_update = f'''
      <EspDrillNavBar 
        currentDay="{current_day_path}" 
        nextDrill="{next_drill_path}" 
        disablePrevious={{ {exercise['Ejercicio #'] == '1'} }} 
        disableNext={{{False}}}
      />
    '''
            updated_content = re.sub(nav_bar_pattern, nav_bar_update, updated_content)

            # Write changes if needed
            if updated_content != content:
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(updated_content)
                self.report['updated_files'] += 1
            else:
                self.report['unchanged'].append(str(file_path))

            return True

        except Exception as e:
            self.report['errors'].append(f"Error updating {file_path}: {str(e)}")
            return False

    def generate_report(self):
        """Generate a detailed report of the update process."""
        print("\n=== UPDATE REPORT ===")
        print(f"Total files: {self.report['total_files']}")
        print(f"Files updated: {self.report['updated_files']}")
        print(f"Files unchanged: {len(self.report['unchanged'])}")
        print(f"Errors: {len(self.report['errors'])}")
        print(f"Mismatches found: {len(self.report['mismatches'])}")

        if self.report['errors']:
            print("\n=== ERRORS ===")
            for error in self.report['errors']:
                print(f"- {error}")

        if self.report['mismatches']:
            print("\n=== MISMATCHES ===")
            for mismatch in self.report['mismatches']:
                print(f"\nFile: {mismatch['file']}")
                print(f"Field: {mismatch['field']}")
                print(f"Current: {mismatch['current']}")
                print(f"Expected: {mismatch['expected']}")

        if self.report['unchanged']:
            print("\n=== UNCHANGED FILES ===")
            for file in self.report['unchanged']:
                print(f"- {file}")

    def run(self):
        """Run the update process."""
        if not self.load_csv():
            print("Failed to load CSV file. Aborting.")
            return

        if not self.verify_file_structure():
            print("File structure verification failed. Aborting.")
            return

        print("\nStarting update process...")
        
        for exercise in tqdm(self.exercises, desc="Updating exercises", unit="file"):
            week = exercise['Semana']
            day = exercise['Día']
            number = exercise['Ejercicio #']
            
            # Create full path with proper week/day numbering
            week_num = week.split()[1]  # Extract the number from "Semana X"
            day_num = day.split()[1]    # Extract the number from "Día X"
            
            file_path = self.base_path / f"semana{week_num}" / f"dia{day_num}" / f"ejercicio{number}.js"
            
            if file_path.exists():
                self.update_file(file_path, exercise)

        self.generate_report()

if __name__ == "__main__":
    updater = ExerciseUpdater()
    updater.run()
